### ISSUE TITLE
Surface Book with Performance Base: nvme and wifi is preventing PC from resuming from s2idle

### ISSUE THREAD

**Issue by kitakar5525** [jakeday/linux-surface#123]
s2idle (S0ix state), formerly called as Suspend-to-Freeze. Actually, some devices with Linux seem to go well with Connected Standby (Modern Standby) when s2idle is selected as a suspend mode (not traditional Suspend-to-Ram which is not supported even in Windows, though).

For example:

- [Dell XPS 13 2-in-1 (9365)](https://wiki.archlinux.org/index.php/Dell_XPS_13_2-in-1_(9365))
- [Lenovo ThinkPad X1 Carbon (Gen 6)](https://wiki.archlinux.org/index.php/Lenovo_ThinkPad_X1_Carbon_(Gen_6))

But my PC fails to resume from s2idle **always** when I attempt to resume **for the second time** (oddly, first time resume is ok).

## Tested environment

1. Surface Book with Performance Base i7/512GB/16GB/dGPU
2. Arch Linux with this repo's kernel (4.15.12-1) and .conf files
  (Also tested on Ubuntu 17.10.1 Live USB with stock kernel and same result)
3. SSD model: THNSN5512GPU7 TOSHIBA
4. wifi model: Marvell Technology Group Ltd. 88W8897 [AVASTAR] 802.11ac Wireless

## How to reproduce

1\. Make sure s2idle is selected as a suspend mode:

```bash
$ cat /sys/power/mem_sleep
[s2idle]
```

(`deep` will also appears if a device is support traditional Suspend-to-Ram, but Surface is not.)

2\. First s2idle, can correctly resume from this state (you can go to s2idle in the ordinary way: `systemctl suspend` and can resume by lid-open)

3\. Second s2idle, and never to resume again

When `no_console_suspend` and `ignore_loglevel` kernel parameter is set, after second resume from s2idle on console, I get messages like this:

![img_20180328_005955_cropped_2048](https://user-images.githubusercontent.com/34676735/38164867-01019cf6-3546-11e8-860b-91b088e796d4.jpg)

It says

```bash
pci 0000:03:00.0: Refused to change power state, currently in D3
nvme 0000:02:00.0: Refused to change power state, currently in D3
```

(pci 0000:03:00.0 is wifi device)

And don't respond to any input anymore.

## Workaround?

```bash
pre/*)
  echo 1 > "/sys/bus/pci/devices/0000:03:00.0/remove"
post/*)
  echo 1 > /sys/bus/pci/rescan
```

in your `/lib/systemd/system-sleep/hibernate.sh` can resolve wifi one.

But nvme will not go the same way. If I do, error storm occurs.

So, I need to set `pcie_port_pm=off` kernel parameter to resume from s2idle for second time and more. However, this may consume more power(?) It consumes 4-5% per hour on s2idle.

Can anyone reproduce this issue? How to change power state of nvme from D3 state?

**Comment by kitakar5525** [comment]
Hi! I can now correctly resume from s2idle more than second time with these commands!:

```bash
echo 0 > "/sys/bus/pci/devices/0000:02:00.0/d3cold_allowed" # nvme
echo 0 > "/sys/bus/pci/devices/0000:03:00.0/d3cold_allowed" # wifi
```

Now no need to set `pcie_port_pm=off` kernel parameter and `echo 1 > "/sys/bus/pci/devices/0000:03:00.0/remove"`, `echo 1 > /sys/bus/pci/rescan` in /lib/systemd/system-sleep/hibernate.sh.

**Comment by mlabieniec** [comment]
@kitakar5525 where are you putting these commands? ... in hibernate.sh -> pre/post ?

**Comment by kitakar5525** [comment]
I put in hibernate.sh -> pre.

**Comment by jakeday** [comment]
@kitakar5525 I'll add it to the script, but I may want to set this in the kernel in the future. Thanks!

**Comment by mlabieniec** [comment]
it doesn't seem to work for me ... well in general `systemctl suspend` doesn't seem to work with or without it...i get a hissing/popping sound on the speakers then no wifi etc when it comes back up a few seconds later. Then my display starts wiggin out

**Comment by jakeday** [comment]
Updated scripts and added patch to kernel.

**Comment by kitakar5525** [comment]
@mlabieniec Setting `no_console_suspend`, `ignore_loglevel` and `systemd.unit=multi-user.target` (system will boot to console) kernel parameters and attempting to suspend may provide you interesting information while your display is still readable.

Because you are using SB2, maybe there are another issues preventing you from suspending the system. I think the cause can differ by models.


@jakeday Just for your information, I noticed only disabling d3cold for nvme is enough (at least for my Performance Base) to resume from s2idle.

```bash
echo 0 > "/sys/bus/pci/devices/0000:02:00.0/d3cold_allowed" # for SB Performance Base nvme
```

Then, I see this message `pci 0000:03:00.0: Refused to change power state, currently in D3` in `dmesg` again each time it resumes, but wifi is still usable each resume from s2idle suspend.

Also see this message `nvme 0000:02:00.0: invalid power transition (from state 4 to 3)` each resume.

I don't know how harmful these message is, though.

**Comment by kitakar5525** [comment]
Sorry to comment on an old closed issue, but I want to share my update.

I wrote my findings here:
- [NVMe SSD not working after s2idle](https://github.com/kitakar5525/note-linux-on-surface-book-1#nvme-ssd-not-working-after-s2idle)

#### summary

I made two patches to fix suspend by kernel side. These patches fix resume from suspend by not allow that NVMe disk to enter D3 state and then, improve power consumption when we disable D3 state on the NVMe during suspend.

- Prevent the NVMe disk from entering D3 state
[prevent-nvme-from-entering-D3.patch](https://github.com/kitakar5525/arch-kernel-linux-lts419-surface/blob/078de5812d240c537b44d9698c91a0cabc55d12b/4416-s2idle-01-5525-prevent-nvme-from-entering-D3.patch)
- Because we disable D3 state for that disk, allow its APST to do power management during suspend
[nvme-add-quirk-to-not-call-disable-function-when-suspending-for-powersaving-419.patch](https://github.com/kitakar5525/arch-kernel-linux-lts419-surface/blob/078de5812d240c537b44d9698c91a0cabc55d12b/4416-s2idle-02-5525-nvme-add-quirk-to-not-call-disable-function-when-suspending-for-powersaving-419.patch)

**Comment by kitakar5525** [comment]
Just sharing the findings, I don't make a PR because I'm not sure I'm doing the right thing🤔

**Comment by kitakar5525** [comment]
I announce here there is an update to 0002-suspend.patch of NVMe part.

### Update patch set for issue "NVMe breaks suspend (s2idle)" and "SSD heating while suspend"

For 5.2
- (Revert the patch 0000-jakeday/0002-suspend.patch of nvme part first
to apply this patch set)
- [nvme: export get and set features · torvalds/linux@1a87ee6](https://github.com/torvalds/linux/commit/1a87ee657c530bb2f3e39e4ac184d48f5f959cda)
- [nvme-pci: use host managed power state for suspend · torvalds/linux@d916b1b](https://github.com/torvalds/linux/commit/d916b1be94b6dc8d293abed2451f3062f6af7551#diff-bc4c090f021c046a7d256a3fcf86b7da)

For 4.19, this patch is also needed
- [nvme-pci: Sync queues on reset · torvalds/linux@d6135c3](https://github.com/torvalds/linux/commit/d6135c3a1ec0cddda7b8b8e1b5b4abeeafd98289#diff-bc4c090f021c046a7d256a3fcf86b7da)

Patch set description:
- Originally, meant to reduce NVMe disk power consumption
while suspend (s2idle) on some NVMe disks.
- As a result, also fixes resuming from s2idle on some devices
- Will be hopefully merged into Linux 5.3
When it arrived, remove this patch set.

Especially, on some Surface devices, for example, my
"Surface Book with Performance Base", which uses TOSHIBA NVMe disk:
```
THNSN5512GPU7 TOSHIBA
Toshiba America Info Systems NVMe Controller [1179:010f]
```
This patch set fixes:
- NVMe SSD heating while s2idle on some NVMe devices
- unable to wakeup from s2idle for second attempt or later on some NVMe devices

Changes from current 0002-suspend.patch:
- It enables on all NVMe disks which supports host managed power state
instead of generic PCI power settings
- whereas current patch enables on only SK Hynix 0x1527 disks and Toshiba 0x010f
disks only

https://github.com/jakeday/linux-surface/blob/3d0abed6c461fd269694b66b9bb6372be230fa20/patches/5.1/0002-suspend.patch#L75-L78




References
- [196907 – [Regression] s2idle does not work with PC300 NVMe SK hynix 512GB - Dell XPS 13 9360](https://bugzilla.kernel.org/show_bug.cgi?id=196907)
- [199689 – s2idle does not work in Dell XPS 9370](https://bugzilla.kernel.org/show_bug.cgi?id=199689)

**Comment by kitakar5525** [comment]
The updated patch has been merged into [qzed:v5.2](https://github.com/qzed/linux-surface/tree/v5.2). It will probably be merged into the jakeday repository later.
